core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteable.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+/**
+ * A {@link Writeable} object identified by its name nad its super type.
+ * To be used for serializable objects that share the same parent class. When reading them, their name tells
+ * which specific instance of object need to be created out of the ones that extend the common parent class.
+ *
+ * @param <C> the super type of the object that gets read/written
+ * @param <T> the specific type of the object that gets read/written
+ */
+public interface NamedWriteable<C, T extends C> extends Writeable<T> {
+
+    /**
+     * Returns the name of the writeable object, needs to be unique for all the objects that share the same super type
+     */
+    String getWriteableName();
+
+    /**
+     * Returns the parent class of the writeable object
+     */
+    Class<C> getWriteableSuperType();
```

Comment:
I think for now lets just go with `getName` and remove the `Class<C>` here we can resolve collisions later